File: bridge_app/lib/services/event_dedupe.dart
/// Capped in-memory deduplication store.
/// Oldest entries are evicted when [maxSize] is exceeded.
class EventDedupe {
  final int maxSize;
  final Set<String> _ids = {};
  final List<String> _queue = [];

  EventDedupe({this.maxSize = 200});

  bool has(String id) => _ids.contains(id);

  void add(String id) {
    if (_ids.contains(id)) return;
    if (_queue.length >= maxSize) {
      final oldest = _queue.removeAt(0);
      _ids.remove(oldest);
    }
    _ids.add(id);
    _queue.add(id);
  }
}

File: bridge-agent/electron/services/EventDedupe.ts
/**
 * In-memory deduplication store capped at `max` entries.
 * Uses insertion-order eviction (oldest dropped first) to bound memory.
 */
export class EventDedupe {
  private readonly _ids = new Set<string>()
  private readonly _queue: string[] = []
  private readonly _max: number

  constructor(max = 200) {
    this._max = max
  }

  has(id: string): boolean {
    return this._ids.has(id)
  }

  add(id: string): void {
    if (this._ids.has(id)) return
    if (this._queue.length >= this._max) {
      const oldest = this._queue.shift()!
      this._ids.delete(oldest)
    }
    this._ids.add(id)
    this._queue.push(id)
  }
}
